File: fix-canonicals.js
const fs = require('fs');
const path = require('path');

function processDir(dir, basePath) {
    const entries = fs.readdirSync(dir, { withFileTypes: true });
    
    for (const entry of entries) {
        const fullPath = path.join(dir, entry.name);
        if (entry.isDirectory()) {
            if (!entry.name.startsWith('(') && !entry.name.startsWith('_')) {
                processDir(fullPath, basePath + '/' + entry.name);
            } else {
                processDir(fullPath, basePath);
            }
        } else if (entry.name === 'page.tsx') {
            let content = fs.readFileSync(fullPath, 'utf8');
            let routePath = basePath || '/';
            
            // Handle dynamic routes like [slug]
            // For SEO canonicals, a dynamic route should ideally generate its canonical in generateMetadata, 
            // but for static pages we can just inject alternates.
            
            if (content.includes('export const metadata')) {
                if (!content.includes('alternates: {') && !content.includes('canonical:')) {
                    content = content.replace(
                        /export const metadata(: Metadata)? = \{/, 
                        xport const metadata = {\n  alternates: { canonical: '' },
                    );
                    fs.writeFileSync(fullPath, content);
                    console.log('Updated: ' + fullPath + ' -> ' + routePath);
                }
            }
        }
    }
}

processDir(path.join(__dirname, 'src', 'app'), '');

File: fix_canonicals.py
import os
import re

app_dir = os.path.join('src', 'app')

def process_dir(directory, base_path):
    for entry in os.scandir(directory):
        if entry.is_dir():
            if not entry.name.startswith('(') and not entry.name.startswith('_'):
                process_dir(entry.path, base_path + '/' + entry.name)
            else:
                process_dir(entry.path, base_path)
        elif entry.name == 'page.tsx':
            with open(entry.path, 'r', encoding='utf-8') as f:
                content = f.read()
                
            route_path = base_path if base_path else '/'
            
            if 'export const metadata' in content and 'alternates: {' not in content:
                content = re.sub(
                    r'export const metadata(: Metadata)? = \{',
                    f"export const metadata\\1 = {{\\n  alternates: {{ canonical: '{route_path}' }},",
                    content,
                    count=1
                )
                with open(entry.path, 'w', encoding='utf-8') as f:
                    f.write(content)
                print(f"Updated: {entry.path} -> {route_path}")

process_dir(app_dir, '')
